File: content/snippets/go/design_patterns_composite.go
package main

import "fmt"

type Node interface {
	TotalSize() int
}

type File struct {
	size int
}

func (f File) TotalSize() int { return f.size }

type Folder struct {
	children []Node
}

func (f *Folder) Add(child Node) {
	f.children = append(f.children, child)
}

func (f Folder) TotalSize() int {
	total := 0
	for _, child := range f.children {
		total += child.TotalSize()
	}
	return total
}

func main() {
	root := &Folder{}
	root.Add(File{size: 5})
	sub := &Folder{}
	sub.Add(File{size: 7})
	sub.Add(File{size: 3})
	root.Add(sub)
	fmt.Println(root.TotalSize())
}

File: content/snippets/cpp/design_patterns_composite.cpp
#include <iostream>
#include <memory>
#include <numeric>
#include <vector>

struct Node {
    virtual ~Node() = default;
    virtual int totalSize() const = 0;
};

struct File : Node {
    explicit File(int size) : size_(size) {}
    int totalSize() const override { return size_; }
    int size_;
};

class Folder : public Node {
public:
    void add(std::unique_ptr<Node> child) {
        children_.push_back(std::move(child));
    }

    int totalSize() const override {
        int sum = 0;
        for (const auto& child : children_) {
            sum += child->totalSize();
        }
        return sum;
    }

private:
    std::vector<std::unique_ptr<Node>> children_;
};

int main() {
    auto root = std::make_unique<Folder>();
    root->add(std::make_unique<File>(5));
    auto sub = std::make_unique<Folder>();
    sub->add(std::make_unique<File>(7));
    sub->add(std::make_unique<File>(3));
    root->add(std::move(sub));
    std::cout << root->totalSize() << '\n';
}
